src/Core/Domain/Product/Image/QueryResult/Shop/ShopProductImages.php

```diff
@@ -0,0 +1,71 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Domain\Product\Image\QueryResult\Shop;
+
+class ShopProductImages
+{
+    /**
+     * @var ProductImage[]
+     */
+    private $productImages = [];
+
+    /**
+     * @var int
+     */
+    private $shopId;
+
+    public function __construct(int $shopId)
+    {
+        $this->shopId = $shopId;
+    }
+
+    /**
+     * @param ProductImage $productImage
+     */
+    public function addProductImage(ProductImage $productImage): void
```

Comment:
Since this is a QueryResult I think it would be preferable to have it as immutable as possible This method is convenient while you are building your object I know that But once it's returned I don't think we want it to be called and the list to be modified

So maybe the list of images should only be passed as the constructor so that it never changes afterwards, it just means you'll need to change the way it's built a little (like store the list per shops in a temp array, and construct the object after you've gone through the whole flatten result)

Reply:
You're right, I was about to do it but by laziness (I'm not proud) I didn't do it. Thank's for the backup :)